tensorflow/core/kernels/scatter_op.h

```diff
@@ -0,0 +1,75 @@
+/* Copyright 2015 Google Inc. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef TENSORFLOW_KERNELS_SCATTER_OP_H_
+#define TENSORFLOW_KERNELS_SCATTER_OP_H_
+
+// Functor definitions for Scatter ops, must be compilable by nvcc.
+
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+#include "tensorflow/core/framework/tensor_types.h"
+
+namespace tensorflow {
+
+class OpKernelContext;
+
+namespace scatter_op {
+
+enum class UpdateOp { ASSIGN, ADD, SUB };
+
+template <UpdateOp Op>
+struct Assign {};
+template <>
+struct Assign<UpdateOp::ASSIGN> {
+  template <typename Params, typename Update>
+  EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE
+  static void Run(Params&& p, const Update& u) {
+    p = u;
+  }
+};
+template <>
+struct Assign<UpdateOp::ADD> {
+  template <typename Params, typename Update>
+  EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE
+  static void Run(Params&& p, const Update& u) {
+    p += u;
```

Comment:
I do understand that `p` in `Params&& p` is really a [universal reference](https://isocpp.org/blog/2012/11/universal-references-in-c11-scott-meyers) and not an rvalue reference. However, it's just not common usage in rest of TF, so I would request you just use a pointer.